pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -161,20 +154,55 @@ public IdealState rebalanceIdealState(IdealState idealState, TableConfig tableCo
       targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
     }
 
-    // always rebalance serving segments
-    rebalanceServingSegments(idealState, tableConfig, targetNumReplicas);
-
     // update if not dryRun
     boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
     if (!dryRun) {
       LOGGER.info("Updating ideal state for table {}", tableNameWithType);
-      updateIdealState(tableNameWithType, targetNumReplicas, idealState.getRecord().getMapFields());
+      rebalanceAndUpdateIdealState(idealState, tableConfig, targetNumReplicas, rebalanceUserConfig,
+          newPartitionAssignment);
     } else {
       LOGGER.info("Dry run. Skip writing ideal state");
+      rebalanceIdealState(idealState, tableConfig, targetNumReplicas, rebalanceUserConfig, newPartitionAssignment);
     }
     return idealState;
   }
 
+  private void rebalanceAndUpdateIdealState(final IdealState rebalanceIdealState, final TableConfig tableConfig,
+      final int targetNumReplicas, final Configuration rebalanceUserConfig,
+      final PartitionAssignment newPartitionAssignment) {
+
+    final Function<IdealState, IdealState> updaterFunction = new Function<IdealState, IdealState>() {
+      @Nullable
+      @Override
+      public IdealState apply(@Nullable IdealState idealState) {
+        rebalanceIdealState(rebalanceIdealState, tableConfig, targetNumReplicas, rebalanceUserConfig,
+            newPartitionAssignment);
+        Map<String, Map<String, String>> mapFields = rebalanceIdealState.getRecord().getMapFields();
+        for (Map.Entry<String, Map<String, String>> entry : mapFields.entrySet()) {
+          idealState.setInstanceStateMap(entry.getKey(), entry.getValue());
+        }
+        idealState.setReplicas(Integer.toString(targetNumReplicas));
+        return idealState;
+      }
+    };
+    HelixHelper.updateIdealState(_helixManager, tableConfig.getTableName(), updaterFunction,
+        RetryPolicies.exponentialBackoffRetryPolicy(5, 1000, 2.0f));
+  }
+
+  private void rebalanceIdealState(IdealState idealState, TableConfig tableConfig, int targetNumReplicas,
```

Comment:
What if we keep `BaseRebalanceSegmentStrategy` and make this `rebalanceIdealState` method as `abtract` method?

Reply:
We already have contract for rebalanceIdealState in the interface. I thought it's better to stick to that one, and let the implementation take care of sub methods. An implementation could want to just fit logic in the top method right?